sphinx_autodoc_typehints.py

```diff
@@ -189,6 +190,115 @@ def process_signature(app, what: str, name: str, obj, options, signature, return
     return signature.format_args().replace('\\', '\\\\'), None
 
 
+def get_all_type_hints(obj, name):
+    rv = {}
+
+    try:
+        rv = get_type_hints(obj)
+    except (AttributeError, TypeError):
+        # Introspecting a slot wrapper will raise TypeError
+        pass
+    except NameError as exc:
+        logger.warning('Cannot resolve forward reference in type annotations of "%s": %s',
+                       name, exc)
+        rv = obj.__annotations__
+
+    if rv:
+        return rv
+
+    rv = backfill_type_hints(obj, name)
+
+    try:
+        obj.__annotations__ = rv
+    except AttributeError:
+        return rv
+
+    try:
+        rv = get_type_hints(obj)
+    except (AttributeError, TypeError):
+        pass
+    except NameError as exc:
+        logger.warning('Cannot resolve forward reference in type annotations of "%s": %s',
+                       name, exc)
+        rv = obj.__annotations__
+
+    return rv
+
+
+def backfill_type_hints(obj, name):
+    rv = {}
+
+    try:
+        import typed_ast.ast3 as ast
+    except ImportError:
+        return rv
+
+    def _one_child(module):
+        children = list(ast.iter_child_nodes(module))
+
+        if len(children) != 1:
+            logger.warning(
```

Comment:
Why this check? This basically disallows doing:
```
def magic(router, **options):
# type: (str, Any) -> Callable[[Any], Any]
```

Reply:
Is this indentation intended? I ran into this warning a bunch of times but didn't investigate which items produce two ast nodes. I am also not sure what to do with two ast nodes

Reply:
just select the first function definition? (type ignore elements should be definitely filtered out) - with the current implementation if the body has type ignores in it this will fail it

Reply:
And what about `@override`?

Reply:
can probably come up with something sane 🤔

Reply:
As said, this doesn't handle all cases, but bailing out seemed safer than selecting a random node without knowledge of why it yielded multiple.

decorators seemed to work for me, but it would make sense to get multiple nodes for that.

Reply:
created https://github.com/agronholm/sphinx-autodoc-typehints/pull/100/files that adds non inlined type comment support 😃